<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Power Analysis</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">differentialDispersion</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="progress.html">Progress</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/awchen55/differentialDispersion">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Power Analysis</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-12-05
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 5
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 2
</p>
<p>
<strong>Knit directory:</strong> <code>differentialDispersion/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240625code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240625)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240625code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240625)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongabsolute">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>File paths:</strong> absolute </a>
</p>
</div>
<div id="strongFilepathsstrongabsolute" class="panel-collapse collapse">
<div class="panel-body">
<p>
Using absolute paths to the files within your workflowr project makes it
difficult for you and others to run your code on a different machine.
Change the absolute path(s) below to the suggested relative path(s) to
make your code more reproducible.
</p>
<table class="table table-condensed table-hover">
<thead>
<tr>
<th style="text-align:left;">
absolute
</th>
<th style="text-align:left;">
relative
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
/project2/gilad/awchen55/differentialDispersion/data/simulations/power_analysis
</td>
<td style="text-align:left;">
data/simulations/power_analysis
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomawchen55differentialDispersiontree33f41a2b1e5cc0af9c6ba2d7c4dcc474cc2b8a1ctargetblank33f41a2a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/awchen55/differentialDispersion/tree/33f41a2b1e5cc0af9c6ba2d7c4dcc474cc2b8a1c" target="_blank">33f41a2</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomawchen55differentialDispersiontree33f41a2b1e5cc0af9c6ba2d7c4dcc474cc2b8a1ctargetblank33f41a2a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/awchen55/differentialDispersion/tree/33f41a2b1e5cc0af9c6ba2d7c4dcc474cc2b8a1c" target="_blank">33f41a2</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/cell_count_metrics_by_cell_type/
    Ignored:    analysis/scran_size_factors_debugging.nb.html
    Ignored:    analysis/sim500_analysis.nb.html
    Ignored:    analysis/simulation_cardiomyocytes_analysis.nb.html
    Ignored:    analysis/simulation_power_analysis.nb.html
    Ignored:    code/.ipynb_checkpoints/
    Ignored:    data/.ipynb_checkpoints/
    Ignored:    data/hybrid_lines_scran_normalized_data/

Untracked files:
    Untracked:  analysis/.ipynb_checkpoints/
    Untracked:  analysis/differential_dispersion_lowess_method.ipynb
    Untracked:  analysis/fitting_ash_hybrid_data.Rmd
    Untracked:  analysis/power_analysis.ipynb
    Untracked:  analysis/scran_size_factors_debugging.Rmd
    Untracked:  analysis/sim500_analysis.Rmd
    Untracked:  analysis/simulation_analysis.ipynb
    Untracked:  analysis/simulation_cardiomyocytes_analysis.Rmd
    Untracked:  analysis/simulation_normal_distribution.Rmd
    Untracked:  code/QTL_power_analysis.R
    Untracked:  code/simulation_background_genes_bin_cv_empirical_p_value_method.py
    Untracked:  code/simulation_background_genes_bin_cv_method.py
    Untracked:  code/step3_simulations_nominal_p_value.py
    Untracked:  data/hybrid_lines_cpm_normalized/
    Untracked:  data/hybrid_lines_log_cpm_regressed_out_data/
    Untracked:  data/hybrid_lines_raw_data/
    Untracked:  data/simulations/

Unstaged changes:
    Modified:   .Rprofile
    Modified:   analysis/differential_dispersion_hybrid_line.Rmd
    Modified:   analysis/differential_dispersion_hybrid_line_cpm_analysis.Rmd
    Modified:   analysis/negative_binomial_dispersion_estimation.Rmd
    Modified:   analysis/simulation_null_analysis.Rmd
    Modified:   analysis/simulation_power_analysis.Rmd
    Modified:   code/step1_simulations.py
    Modified:   code/step2_simulations.py
    Modified:   code/step3_simulations.py
    Deleted:    data/chimp_ASE_subset.csv
    Deleted:    data/human_ASE_subset.csv

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/simulation_power_analysis.Rmd</code>) and HTML
(<code>docs/simulation_power_analysis.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/33f41a2b1e5cc0af9c6ba2d7c4dcc474cc2b8a1c/analysis/simulation_power_analysis.Rmd" target="_blank">33f41a2</a>
</td>
<td>
awchen55
</td>
<td>
2024-12-05
</td>
<td>
update power analysis
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/awchen55/differentialDispersion/33f41a2b1e5cc0af9c6ba2d7c4dcc474cc2b8a1c/docs/simulation_power_analysis.html" target="_blank">33f41a2</a>
</td>
<td>
awchen55
</td>
<td>
2024-12-05
</td>
<td>
update power analysis
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/60a5b9dede71eb6a1bf0073f78ac5bb22b51842d/analysis/simulation_power_analysis.Rmd" target="_blank">60a5b9d</a>
</td>
<td>
awchen55
</td>
<td>
2024-12-02
</td>
<td>
add power analysis
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/awchen55/differentialDispersion/60a5b9dede71eb6a1bf0073f78ac5bb22b51842d/docs/simulation_power_analysis.html" target="_blank">60a5b9d</a>
</td>
<td>
awchen55
</td>
<td>
2024-12-02
</td>
<td>
add power analysis
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="power-analysis-details" class="section level2">
<h2><strong>Power Analysis Details</strong></h2>
<p>We perform a power analysis for QTLs using our simulations. We are
asking two main questions. At what effect size do you detect 80% of true
positives in genes that have 1 UMI/cell and 1k cells? At that same
effect size, what about 2 UMI/cell in 500 cells? 0.5 UMI/cell in 2000
cells? Additionally, at the same effect size, what about across
different numbers of individuals?</p>
</div>
<div id="simulations-at-1000-cells-and-1-umicell"
class="section level2">
<h2><strong>Simulations at 1000 cells and 1 UMI/cell</strong></h2>
<p>Under the simulation framework for <span
class="math inline">\(i\)</span> cells and <span
class="math inline">\(j\)</span> genes: <span class="math inline">\(E[X]
= L_i\alpha_j\sigma_j\)</span> and <span class="math inline">\(Var(X)
=L_i\alpha_j\sigma_j + L_i^2 \alpha_j\sigma_j^2\)</span>, where <span
class="math inline">\(L\)</span> is the library size and <span
class="math inline">\(\alpha, \sigma\)</span> are the shape and scale
parameters. To increase the dispersion level we multiply <span
class="math inline">\(\sigma\)</span> by an applied factor. We assume a
fixed library size for all cells. We can set the target of 1 UMI/cell by
setting the expectation equal to 1 for a fixed <span
class="math inline">\(L,\alpha, \text{and } \sigma\)</span>.</p>
<div id="estimating-phi" class="section level3">
<h3><strong><em>Estimating <span
class="math inline">\(\phi\)</span></em></strong></h3>
<p>We use the HDC dataset and look at the cardiomyocyte cell type. We
pool genes that have mean expression of between 0.9 - 1.1 UMI/cell and
fit a Gamma-Poisson model to estimate <span
class="math inline">\(\phi\)</span>.</p>
<pre class="r"><code>library(glmGamPoi)
library(dplyr)

card_path = &quot;/project/gilad/brendan/dispersion/pilot/cHDC_data/cellranger_cluster-mode_trial/analysis/datasets/median_gene_detection/&quot;
#setwd(card_path)


cardiomyocytes_data4000 &lt;- readRDS(paste(card_path, &quot;lane_a_card_downsampled_UMI_matrix_10.rds&quot;, sep = &quot;&quot;))
cardiomyocytes_data4000 &lt;- t(as.matrix(cardiomyocytes_data4000))

# calculate mean expression and remove genes with low expression
cardiomyocytes_data4000_filtered &lt;- cardiomyocytes_data4000[, colSums(cardiomyocytes_data4000 != 0) &gt; 10]

# calculate mean expression
mean_expression4000 &lt;- colMeans(cardiomyocytes_data4000_filtered)</code></pre>
<p>We group the genes within 0.9-1.1 UMI/cell. There are 778 genes.</p>
<pre class="r"><code># get indices for genes within bin of mean expression
mean_idx &lt;- which(between(mean_expression4000,0.9,1.1)) 
grouped_data &lt;- as.vector(cardiomyocytes_data4000_filtered[,mean_idx])
length(mean_idx)</code></pre>
<pre><code>[1] 778</code></pre>
<p>Of these genes we also get an estimate of the average library size
and use this as the fixed library size in the simulations.</p>
<pre class="r"><code>avg_li &lt;- mean(colSums(cardiomyocytes_data4000_filtered[,mean_idx]))
avg_li</code></pre>
<pre><code>[1] 4064.29</code></pre>
<p>We now fit the Gamma-Poisson to the pooled data and estimate <span
class="math inline">\(\phi\)</span>.</p>
<pre class="r"><code># fit gamma-poisson distribution
fit_gp &lt;- glm_gp(grouped_data, size_factors=1)
sum_gp &lt;- summary(fit_gp)
phi &lt;- sum_gp$overdispersions
phi</code></pre>
<pre><code>[1] 0.8917347</code></pre>
</div>
<div id="set-l-alpha-sigma" class="section level3">
<h3><strong><em>Set <span class="math inline">\(L, \alpha,
\sigma\)</span></em></strong></h3>
<p>We estimate the Gamma-Poisson values using
<strong>glmGamPoi</strong>: <span class="math display">\[
E[X] = \mu \hspace{20 mm} Var[X] = \mu + \mu^2\phi
\]</span></p>
<p>Under the simulation framework: <span class="math display">\[
E[X] = L_i\alpha_j\sigma_j \hspace{20 mm} Var(X) =L_i\alpha_j\sigma_j +
L_i^2 \alpha_j\sigma_j^2
\]</span></p>
<p>Setting the expectation and variance equations equal to each other,
we get: <span class="math display">\[
\sigma = \frac{\mu\phi}{L} \hspace{20 mm} \alpha = \frac{1}{\phi}
\]</span></p>
<p>We know that <span class="math inline">\(\mu = 1\)</span> since we
are targeting 1 UMI/cell. We multiply <span
class="math inline">\(L_i\alpha_j\sigma_j\)</span> to double check that
it equals 1.</p>
<pre class="r"><code>mu &lt;- 1

sigma &lt;- mu*phi/avg_li
alpha &lt;- 1/phi

sigma*alpha*avg_li</code></pre>
<pre><code>[1] 1</code></pre>
</div>
<div id="simulate-at-varying-dispersion-levels" class="section level3">
<h3><strong><em>Simulate at varying dispersion levels</em></strong></h3>
<p>For a fixed mean of 1 UMI/cell we want to test different effect
sizes. We assess different effect sizes by varying the applied factor
from: 1,1.1,1.2,…,2. We will do this and test across different
individuals per group: 1, 2, 3, … , 25. Since we are testing for 1000
cells/individual we can pool all cells for individuals. For example, 25
individuals/group would be 25,000 cells in the null variance group and
35,000 cells in the effect size group for a total of 50 individuals.</p>
<pre class="r"><code>sim_gene &lt;- function(n,alpha,sigma,Li, sigma_multiplier=1){

  # Simulate a single gene with following parameters:
  # n:                number of cells
  # alpha:            alpha parameter
  # sigma:            sigma parameter
  # Li:               library size
  # sigma_multiplier: factor to increase variance
 
  # Returns expression values for single gene in n cells
  
  # initialize Y
  Y &lt;- rep(NA,n) 
  
  # calculate latent expression using gamma distribution
  # ensure that simulation has mean within 10% of target
  # perc_diff = 1
  # while(perc_diff &gt; 0.05){
  lambdaj &lt;- rgamma(n,shape = alpha/sigma_multiplier,scale = sigma*sigma_multiplier)
  # perc_diff = (alpha*sigma - mean(lambdaj))/alpha*sigma
  # }

  
  #lambdaj &lt;- rgamma(n,shape = alpha/sigma_multiplier,scale = sigma*sigma_multiplier)
  
  # for each cell use the latent expression and library size to simulate expression
  for(i in 1:n){
  
    Y[i] &lt;- rpois(1,Li*lambdaj[i])
  }
  
  return(Y)
}</code></pre>
<p>We perform the simulation for 1 individual/group and check the mean
UMI.</p>
<pre class="r"><code>set.seed(11222024)
varying_factor &lt;- seq(1, 2, by = 0.1)

gene_sim_matrix = c()
for(i in 1:length(varying_factor)){
  gene_sim &lt;- sim_gene(1000,alpha,sigma,avg_li,sigma_multiplier=varying_factor[i])
  gene_sim_matrix &lt;- cbind(gene_sim_matrix,gene_sim)
}
colnames(gene_sim_matrix) &lt;- make.names(varying_factor)
colMeans(gene_sim_matrix)</code></pre>
<pre><code>   X1  X1.1  X1.2  X1.3  X1.4  X1.5  X1.6  X1.7  X1.8  X1.9    X2 
0.962 0.972 0.962 1.007 1.042 1.012 0.970 1.114 0.905 0.906 1.078 </code></pre>
<p>We now save the matrix data for each number of individuals.</p>
<pre class="r"><code>set.seed(11222024)

path = &quot;/project2/gilad/awchen55/differentialDispersion/data/simulations/power_analysis&quot;

varying_factor &lt;- seq(1, 2, by = 0.1)
num_indv &lt;- c(1,seq(5, 50, by = 5))

for(j in 1:length(num_indv)){
  gene_sim_matrix = c()
  num_indv_grp = 1000*num_indv[j]
  
  for(i in 1:length(varying_factor)){
    gene_sim &lt;- sim_gene(num_indv_grp,alpha,sigma,avg_li,sigma_multiplier=varying_factor[i])
    gene_sim_matrix &lt;- cbind(gene_sim_matrix,gene_sim)
    
  }
  colnames(gene_sim_matrix) &lt;- make.names(varying_factor)
  write.csv(as.data.frame(gene_sim_matrix), file = file.path(path,paste(&quot;power_analysis_indv_&quot;,num_indv[j],&quot;.csv&quot;,sep=&quot;&quot;)), row.names = FALSE)
}</code></pre>
<pre class="r"><code>path = &quot;/project2/gilad/awchen55/differentialDispersion/data/simulations/power_analysis&quot;
head(read.csv(file.path(path, &quot;power_analysis_indv_50.csv&quot;)))</code></pre>
<pre><code>  X1 X1.1 X1.2 X1.3 X1.4 X1.5 X1.6 X1.7 X1.8 X1.9 X2
1  0    0    0    0    0    0    0    5    2    2  1
2  2    0    0    0    3    1    7    0    0    7  0
3  0    2    1    0    2    0    0    3    1    0  1
4  1    0    3    3    0    0    0    1    0    0  2
5  0    4    2    2    5    3    0    0    1    0  0
6  0    0    0    1    0    0    1    0    1    0  0</code></pre>
</div>
</div>
<div id="estimating-variance" class="section level2">
<h2><strong>Estimating Variance</strong></h2>
<p>We use the following equation to estimate variance by assuming a
multiple hypergeometric model for the data. <a
href="https://www.cell.com/cell/fulltext/S0092-8674%2824%2901144-9">Kim
et al.</a> uses method of moments to estimate the variance.</p>
<p><span class="math display">\[
\hat{\sigma}_{g,HG}^2 = \frac{1}{n_{cells}}\sum_c \frac{Y^2_{cg} -
Y_{cg}(1-q)}{N^2_cq^2- N_cq(1-q)} - \left(
\frac{1}{n_{cells}}\sum_c\frac{Y_{cg}}{N_cq} \right)^2
\]</span> Where for <span class="math inline">\(c\)</span> cells and
<span class="math inline">\(g\)</span> genes, <span
class="math inline">\(Y_{cg}\)</span> is the observed transcript count,
<span class="math inline">\(N_c\)</span> is the true total transcripts,
<span class="math inline">\(n_{cells}\)</span> is the number of cells,
and <span class="math inline">\(q\)</span> is the sampling efficiency.
<span class="math inline">\(N_cq\)</span> is the library size.</p>
</div>
<div id="determine-significance" class="section level2">
<h2><strong>Determine Significance</strong></h2>
<p>We are assuming a two-sample problem between groups A and B, where A
is the group with no effect size and B is the group with an effect size.
Let <span class="math inline">\(\Delta t\)</span> be the observed
difference between groups A and B, where <span
class="math inline">\(t\)</span> is the statistic of interest. In our
case, we are interested in differences in variance for a fixed mean. We
perform bootstrapping for <span class="math inline">\(B\)</span>
iterations to create a sampling distribution of <span
class="math inline">\(\Delta t_1,\dots,\Delta t_B\)</span>. To create
the null distribution we subtract <span class="math inline">\(\Delta
t\)</span> from <span class="math inline">\(\Delta t_1,\dots,\Delta
t_B\)</span> to calculate <span class="math inline">\(\Delta
t_1^*,\dots,\Delta t_B^*\)</span>. This is the same approach used by Kim
et al. We then compute the achieved significance level (ASL) as
described by Efron and Tibshirani:</p>
<p><span class="math display">\[
\text{ASL} =
\begin{cases}
\frac{1}{B}\sum^B_{i=1}1(\Delta t_i^* \geq\Delta t) , &amp; \text{if }
\Delta t \geq 0 \\
    \frac{1}{B} \sum_{i=1}^B 1(\Delta t_i^* \leq\Delta t), &amp;
\text{if } \Delta t &lt; 0
\end{cases}
\
\]</span></p>
<pre class="r"><code>sigma_sq_est &lt;- function(df,n_cells,q,N_c){
  sigma_sq_term1 = (df^2 - df*(1-q))/(N_c^2*q^2 - N_c*q*(1-q))
  sigma_sq_term1 = colSums(sigma_sq_term1)/n_cells

  sigma_sq_term2 = df/(N_c*q)
  sigma_sq_term2 = (colSums(sigma_sq_term2)/n_cells)^2

  sigma_sq = sigma_sq_term1 - sigma_sq_term2
  
  
  return(sigma_sq)
}</code></pre>
</div>
<div id="run-power-analysis" class="section level2">
<h2><strong>Run Power Analysis</strong></h2>
<p>We run the power analysis by simulating 1,000 times for <span
class="math inline">\(B=1,000\)</span> bootstrapping iterations. We
assume <span class="math inline">\(q=0.4\)</span>. We test out 100
different effect sizes for the applied factor to the variance, spanned
evenly between 1 and 2.</p>
<pre class="r"><code>n_sims = 100
B = 1000
n_cells = 1000
n_indv = 1
q = 0.4
library_size = 4064.29
N_c = library_size/q

varying_factor &lt;- seq(1, 2, by = 100)
#num_indv &lt;- c(1,seq(5, 50, by = 5))


n_sig &lt;- rep(0,(length(varying_factor)-1))
for(j in 1:n_sims){
  gene_sim_matrix = c()
  num_indv_grp = n_cells*n_indv
  
  for(i in 1:length(varying_factor)){
    gene_sim &lt;- sim_gene(num_indv_grp,alpha,sigma,avg_li,sigma_multiplier=varying_factor[i])
    gene_sim_matrix &lt;- cbind(gene_sim_matrix,gene_sim)
    
  }
  colnames(gene_sim_matrix) &lt;- make.names(varying_factor)
  
  sigma_sq_obs = sigma_sq_est(gene_sim_matrix,num_indv_grp,q,N_c)
  delta_t = sigma_sq_obs[2:length(sigma_sq_obs)] - sigma_sq_obs[1]
  
  
  asl_geq &lt;- rep(0, (length(varying_factor)-1))
  asl_lt &lt;- rep(0, (length(varying_factor)-1))
  for(i in 1:B){
    sample_idx &lt;- sample(1:num_indv_grp,size = num_indv_grp,replace=TRUE)
    sample_df &lt;- gene_sim_matrix[sample_idx,]
    
    sigma_sq_samp = sigma_sq_est(sample_df,num_indv_grp,q,N_c)
    delta_t_boot = sigma_sq_samp[2:length(sigma_sq_samp)] - sigma_sq_samp[1]
    delta_t_star = delta_t_boot - delta_t
    
    asl_geq = asl_geq + as.numeric(delta_t_star &gt;= delta_t)
    asl_lt = asl_lt + as.numeric(delta_t_star &lt;= delta_t)
    
  }
    
  
  asl &lt;-c()
  for(i in 1:length(delta_t)){
    if(delta_t[i] &gt;=0){
      asl &lt;- c(asl,asl_geq[i])
    }
    else{
      asl &lt;- c(asl,asl_lt[i])
    }
  }
  
  asl &lt;- asl/B
  
  n_sig = n_sig + as.numeric(asl&lt;0.05)
}

power_vec &lt;- n_sig/n_sims
names(power_vec) &lt;- varying_factor[2:length(varying_factor)]

path = &quot;/project2/gilad/awchen55/differentialDispersion/data/simulations/power_analysis&quot;

save(power_vec, file = file.path(path,paste(&quot;power_analysis_indv_&quot;,n_indv,&quot;.RData&quot;,sep=&quot;&quot;)))</code></pre>
<div id="power-analysis-across-effect-size" class="section level3">
<h3><strong><em>Power Analysis Across Effect Size</em></strong></h3>
<p>We look at the power based on varying effect size for the number of
individuals.</p>
<pre class="r"><code># Initialize an empty list
data_list &lt;- list()
path = &quot;/project2/gilad/awchen55/differentialDispersion/data/simulations/power_analysis&quot;

for (n_indv in 1:8) {
  # Load the RData file into a temporary environment
  temp_env &lt;- new.env()
  load(file.path(path, paste(&quot;power_analysis_100ef_indv_&quot;, n_indv, &quot;.RData&quot;, sep = &quot;&quot;)), envir = temp_env)
  
  # Extract the power_vec and remove the last element
  powervec1 &lt;- temp_env$power_vec
  powervec1 &lt;- powervec1[-length(powervec1)]  # Remove the last element
  
  # Append the loaded data to the list with an &#39;id&#39; column
  data_list[[n_indv]] &lt;- data.frame(x = as.numeric(names(powervec1)), y = powervec1, id = factor(n_indv))
}

# Combine the data into a single data frame
combined_data &lt;- do.call(rbind, data_list)</code></pre>
<pre class="r"><code>library(ggplot2)
# Colors for the graphs
#colors &lt;- c(&quot;red&quot;, &quot;blue&quot;, &quot;green&quot;)

ggplot(combined_data, aes(x = x, y = y, color = as.factor(id), group = as.factor(id))) +
  geom_line(size = 1.2) + ylim(0,1)+
  labs(title = &quot;Power Analysis Across Effect Size&quot;,
       x = &quot;effect size&quot;, 
       y = &quot;power&quot;,
       color = &quot;Number of Individuals per Group&quot;) +
  #scale_color_manual(values = c(&quot;red&quot;, &quot;blue&quot;, &quot;green&quot;)) +  # Manually assign colors to each group
  theme_minimal() +
  theme(legend.position = &quot;top&quot;) </code></pre>
<p><img src="figure/simulation_power_analysis.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-13-1">
Past versions of unnamed-chunk-13-1.png
</button>
</p>
<div id="fig-unnamed-chunk-13-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/33f41a2b1e5cc0af9c6ba2d7c4dcc474cc2b8a1c/docs/figure/simulation_power_analysis.Rmd/unnamed-chunk-13-1.png" target="_blank">33f41a2</a>
</td>
<td>
awchen55
</td>
<td>
2024-12-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="power-analysis-across-individuals" class="section level3">
<h3><strong><em>Power Analysis Across Individuals</em></strong></h3>
<pre class="r"><code>ef_1.2 &lt;- combined_data[round(combined_data$x, 6) == 1.202020, ]
ef_1.25 &lt;- combined_data[round(combined_data$x, 6) == 1.252525, ]
ef_1.3 &lt;- combined_data[round(combined_data$x, 6) == 1.303030, ]
ef_1.35 &lt;- combined_data[round(combined_data$x, 6) == 1.353535, ]
ef_1.4 &lt;- combined_data[round(combined_data$x, 6) == 1.404040, ]
ef_1.45 &lt;- combined_data[round(combined_data$x, 6) == 1.454545, ]
ef_1.5 &lt;- combined_data[round(combined_data$x, 6) == 1.505051, ]
combined_ef &lt;- rbind(ef_1.2,ef_1.25,ef_1.3,ef_1.35,ef_1.4,ef_1.45,ef_1.5)

ggplot(combined_ef, aes(x = id, y = y, color = as.factor(round(x,2)), group = as.factor(round(x,2)))) +
  geom_line(size = 1.2) + ylim(0, 1) +
  labs(title = &quot;Power Analysis Across Individuals&quot;,
       x = &quot;number of individuals per group&quot;, 
       y = &quot;power&quot;,
       color = &quot;Effect Size&quot;) +
  #scale_color_manual(values = c(&quot;red&quot;, &quot;blue&quot;, &quot;green&quot;)) +  # Manually assign colors to each group
  theme_minimal() +
  theme(legend.position = &quot;top&quot;) </code></pre>
<p><img src="figure/simulation_power_analysis.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.0 (2022-04-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.3.13-el7-x86_64/lib/libopenblas_haswellp-r0.3.13.so

locale:
 [1] LC_CTYPE=en_US.UTF-8 LC_NUMERIC=C         LC_TIME=C           
 [4] LC_COLLATE=C         LC_MONETARY=C        LC_MESSAGES=C       
 [7] LC_PAPER=C           LC_NAME=C            LC_ADDRESS=C        
[10] LC_TELEPHONE=C       LC_MEASUREMENT=C     LC_IDENTIFICATION=C 

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_3.5.1   dplyr_1.1.4     glmGamPoi_1.8.0

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.12                 lattice_0.20-45            
 [3] rprojroot_2.0.3             digest_0.6.29              
 [5] utf8_1.2.2                  R6_2.5.1                   
 [7] GenomeInfoDb_1.32.2         stats4_4.2.0               
 [9] evaluate_0.15               highr_0.9                  
[11] pillar_1.9.0                sparseMatrixStats_1.8.0    
[13] zlibbioc_1.42.0             rlang_1.1.4                
[15] rstudioapi_0.13             whisker_0.4                
[17] jquerylib_0.1.4             S4Vectors_0.34.0           
[19] Matrix_1.6-4                rmarkdown_2.14             
[21] labeling_0.4.2              stringr_1.5.1              
[23] RCurl_1.98-1.7              munsell_0.5.0              
[25] DelayedArray_0.22.0         compiler_4.2.0             
[27] httpuv_1.6.5                xfun_0.30                  
[29] pkgconfig_2.0.3             BiocGenerics_0.42.0        
[31] htmltools_0.5.2             tidyselect_1.2.1           
[33] SummarizedExperiment_1.26.1 tibble_3.2.1               
[35] GenomeInfoDbData_1.2.8      workflowr_1.7.1            
[37] IRanges_2.30.0              matrixStats_0.62.0         
[39] fansi_1.0.3                 withr_3.0.0                
[41] later_1.3.0                 bitops_1.0-7               
[43] grid_4.2.0                  jsonlite_1.8.0             
[45] gtable_0.3.0                lifecycle_1.0.4            
[47] git2r_0.30.1                magrittr_2.0.3             
[49] scales_1.3.0                cli_3.6.3                  
[51] stringi_1.7.6               farver_2.1.0               
[53] XVector_0.36.0              fs_1.5.2                   
[55] promises_1.2.0.1            bslib_0.3.1                
[57] DelayedMatrixStats_1.18.2   vctrs_0.6.5                
[59] generics_0.1.3              tools_4.2.0                
[61] Biobase_2.56.0              glue_1.7.0                 
[63] MatrixGenerics_1.8.0        fastmap_1.1.0              
[65] yaml_2.3.5                  colorspace_2.0-3           
[67] GenomicRanges_1.48.0        knitr_1.39                 
[69] sass_0.4.1                 </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
