<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2024-07-24" />

<title>Poisson and Negative Binomial Simulations</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">differentialDispersion</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="progress.html">Progress</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/awchen55/differentialDispersion">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore"><strong>Poisson and Negative Binomial
Simulations</strong></h1>
<h4 class="date">2024-07-24</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-08-16
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 5
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 2
</p>
<p>
<strong>Knit directory:</strong> <code>differentialDispersion/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240625code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240625)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240625code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240625)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongabsolute">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>File paths:</strong> absolute </a>
</p>
</div>
<div id="strongFilepathsstrongabsolute" class="panel-collapse collapse">
<div class="panel-body">
<p>
Using absolute paths to the files within your workflowr project makes it
difficult for you and others to run your code on a different machine.
Change the absolute path(s) below to the suggested relative path(s) to
make your code more reproducible.
</p>
<table class="table table-condensed table-hover">
<thead>
<tr>
<th style="text-align:left;">
absolute
</th>
<th style="text-align:left;">
relative
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
/project2/gilad/awchen55/differentialDispersion/data/hybrid_lines_raw_data/human.ASE.Rep3.h5Seurat
</td>
<td style="text-align:left;">
data/hybrid_lines_raw_data/human.ASE.Rep3.h5Seurat
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomawchen55differentialDispersiontreef599f8dbcff6e9f62f33737f800434a72bb7b68btargetblankf599f8da">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/awchen55/differentialDispersion/tree/f599f8dbcff6e9f62f33737f800434a72bb7b68b" target="_blank">f599f8d</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomawchen55differentialDispersiontreef599f8dbcff6e9f62f33737f800434a72bb7b68btargetblankf599f8da"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/awchen55/differentialDispersion/tree/f599f8dbcff6e9f62f33737f800434a72bb7b68b" target="_blank">f599f8d</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/cell_count_metrics_by_cell_type/
    Ignored:    analysis/scran_size_factors_debugging.nb.html
    Ignored:    code/.ipynb_checkpoints/
    Ignored:    data/.ipynb_checkpoints/
    Ignored:    data/hybrid_lines_scran_normalized_data/

Untracked files:
    Untracked:  analysis/.ipynb_checkpoints/
    Untracked:  analysis/differential_dispersion_lowess_method.ipynb
    Untracked:  analysis/fitting_ash_hybrid_data.Rmd
    Untracked:  analysis/scran_size_factors_debugging.Rmd
    Untracked:  data/hybrid_lines_cpm_normalized/
    Untracked:  data/hybrid_lines_log_cpm_regressed_out_data/
    Untracked:  data/hybrid_lines_raw_data/

Unstaged changes:
    Modified:   .Rprofile
    Modified:   analysis/differential_dispersion_hybrid_line.Rmd
    Modified:   analysis/differential_dispersion_hybrid_line_cpm_analysis.Rmd
    Modified:   analysis/negative_binomial_dispersion_estimation.Rmd
    Modified:   analysis/simulation_null_analysis.Rmd
    Modified:   code/step1_calculate_cell_count_metrics_lowess_method.py
    Modified:   code/step2_residual_calculations_lowess_values.py
    Modified:   code/step3_bootstrapping_cell_types_lowess_method.py
    Deleted:    data/chimp_ASE_subset.csv
    Deleted:    data/human_ASE_subset.csv

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/simulation_null_analysis.Rmd</code>) and HTML
(<code>docs/simulation_null_analysis.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/f599f8dbcff6e9f62f33737f800434a72bb7b68b/analysis/simulation_null_analysis.Rmd" target="_blank">f599f8d</a>
</td>
<td>
awchen55
</td>
<td>
2024-08-15
</td>
<td>
updated simulation results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/awchen55/differentialDispersion/f599f8dbcff6e9f62f33737f800434a72bb7b68b/docs/simulation_null_analysis.html" target="_blank">f599f8d</a>
</td>
<td>
awchen55
</td>
<td>
2024-08-15
</td>
<td>
updated simulation results
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/23e157da6638082dc5079827509d1f546ebfe4e9/analysis/simulation_null_analysis.Rmd" target="_blank">23e157d</a>
</td>
<td>
awchen55
</td>
<td>
2024-08-13
</td>
<td>
simulation results for disp
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/awchen55/differentialDispersion/23e157da6638082dc5079827509d1f546ebfe4e9/docs/simulation_null_analysis.html" target="_blank">23e157d</a>
</td>
<td>
awchen55
</td>
<td>
2024-08-13
</td>
<td>
simulation results for disp
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/5285b92dc23ad3d7d861d43a80d1527e7da5d279/analysis/simulation_null_analysis.Rmd" target="_blank">5285b92</a>
</td>
<td>
awchen55
</td>
<td>
2024-08-07
</td>
<td>
update simulation results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/awchen55/differentialDispersion/5285b92dc23ad3d7d861d43a80d1527e7da5d279/docs/simulation_null_analysis.html" target="_blank">5285b92</a>
</td>
<td>
awchen55
</td>
<td>
2024-08-07
</td>
<td>
update simulation results
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/38913486e7b6efe14e2f08af0fb394a102a1a3b5/analysis/simulation_null_analysis.Rmd" target="_blank">3891348</a>
</td>
<td>
awchen55
</td>
<td>
2024-08-06
</td>
<td>
update simulation
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/awchen55/differentialDispersion/38913486e7b6efe14e2f08af0fb394a102a1a3b5/docs/simulation_null_analysis.html" target="_blank">3891348</a>
</td>
<td>
awchen55
</td>
<td>
2024-08-06
</td>
<td>
update simulation
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/b3de4f69ff5d8e6fb9ffc0f78c1406b42db466a7/analysis/simulation_null_analysis.Rmd" target="_blank">b3de4f6</a>
</td>
<td>
awchen55
</td>
<td>
2024-08-06
</td>
<td>
update simulation results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/awchen55/differentialDispersion/b3de4f69ff5d8e6fb9ffc0f78c1406b42db466a7/docs/simulation_null_analysis.html" target="_blank">b3de4f6</a>
</td>
<td>
awchen55
</td>
<td>
2024-08-06
</td>
<td>
update simulation results
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/e89b6f835e3619ad650d0de07a22085ff5c2a3ef/analysis/simulation_null_analysis.Rmd" target="_blank">e89b6f8</a>
</td>
<td>
awchen55
</td>
<td>
2024-07-30
</td>
<td>
outline procedure using ash
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/awchen55/differentialDispersion/e89b6f835e3619ad650d0de07a22085ff5c2a3ef/docs/simulation_null_analysis.html" target="_blank">e89b6f8</a>
</td>
<td>
awchen55
</td>
<td>
2024-07-30
</td>
<td>
outline procedure using ash
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/af01093ad7886b2157bbfdabe33ac53401483a09/analysis/simulation_null_analysis.Rmd" target="_blank">af01093</a>
</td>
<td>
awchen55
</td>
<td>
2024-07-30
</td>
<td>
update simulation details
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/awchen55/differentialDispersion/af01093ad7886b2157bbfdabe33ac53401483a09/docs/simulation_null_analysis.html" target="_blank">af01093</a>
</td>
<td>
awchen55
</td>
<td>
2024-07-30
</td>
<td>
update simulation details
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/5eda9a3727cff50da05da8e4bca48f254b492fed/analysis/simulation_null_analysis.Rmd" target="_blank">5eda9a3</a>
</td>
<td>
awchen55
</td>
<td>
2024-07-25
</td>
<td>
added details of simulations
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/awchen55/differentialDispersion/5eda9a3727cff50da05da8e4bca48f254b492fed/docs/simulation_null_analysis.html" target="_blank">5eda9a3</a>
</td>
<td>
awchen55
</td>
<td>
2024-07-25
</td>
<td>
added details of simulations
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/f2f7418058b176bc84d9dc8250b3fd6077cf2201/analysis/simulation_null_analysis.Rmd" target="_blank">f2f7418</a>
</td>
<td>
awchen55
</td>
<td>
2024-07-24
</td>
<td>
add simulation details
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/awchen55/differentialDispersion/f2f7418058b176bc84d9dc8250b3fd6077cf2201/docs/simulation_null_analysis.html" target="_blank">f2f7418</a>
</td>
<td>
awchen55
</td>
<td>
2024-07-24
</td>
<td>
add simulation details
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="simulation-details" class="section level1">
<h1><strong>Simulation Details</strong></h1>
<p>We simulate a matrix <span class="math inline">\(Y \in
\mathbb{R}^{n\times p}\)</span> that contains raw expression values
(scRNA-seq counts) for <span class="math inline">\(p\)</span> genes and
<span class="math inline">\(n\)</span> single cells. We set <span
class="math inline">\(\pi\)</span> as the percent of genes that are
overdispersed. We use a Gamma-Poisson model for each gene and can set
the dispersion level for each simulated gene.</p>
<div id="gamma-poisson-nb-distribution" class="section level2">
<h2><strong>Gamma-Poisson (NB) Distribution</strong></h2>
<div id="model" class="section level3">
<h3><strong><em>Model</em></strong></h3>
<p>We simulate the Negative Binomial distribution using the
Gamma-Poisson distribution, since the Gamma distribution is a conjugate
prior of the Poisson distribution. Under this model,</p>
<p><span class="math display">\[
y_{ij} \sim Pois(L_i \lambda_j)
\]</span></p>
<p><span class="math display">\[
\lambda_j \sim \begin{equation}
  f(\cdot) =
    \begin{cases}
      \pi &amp; Gamma(\mu_j,\phi_j)\\
      (1-\pi) &amp; Gamma(\mu_0,\phi_0) \\
    \end{cases}       
\end{equation}
\]</span></p>
<p>Where the variables for cell <span class="math inline">\(i\)</span>
and gene <span class="math inline">\(j\)</span> are:</p>
<ul>
<li><p><span class="math inline">\(y_{ij}\)</span> is the number of
unique molecular identifier (UMI) for cell <span
class="math inline">\(i\)</span> and gene <span
class="math inline">\(j\)</span></p></li>
<li><p><span class="math inline">\(L_i\)</span> is the library size of
cell <span class="math inline">\(i\)</span></p></li>
<li><p><span class="math inline">\(\lambda_j\)</span> is the latent
expression value of gene <span class="math inline">\(j\)</span></p></li>
<li><p><span class="math inline">\(\pi\)</span> is the percent of genes
that are overdispersed</p></li>
<li><p><span class="math inline">\(\mu_j\)</span> is the mean relative
expression value of <span class="math inline">\(\lambda_j\)</span> for
gene <span class="math inline">\(j\)</span></p></li>
<li><p><span class="math inline">\(\mu_0\)</span> is the null relative
expression value for a gene that has no dispersion</p></li>
<li><p><span class="math inline">\(\phi_j\)</span> is the dispersion
value of <span class="math inline">\(\lambda_j\)</span> for gene <span
class="math inline">\(j\)</span></p></li>
<li><p><span class="math inline">\(\phi_0\)</span> is the null relative
dispersion value for a gene that has no dispersion</p></li>
</ul>
</div>
<div id="density-function" class="section level3">
<h3><strong><em>Density Function</em></strong></h3>
<p>The density function for the Gamma distribution is:</p>
<p><span class="math display">\[
f(y; \mu,\phi) =
\frac{(\phi^{-1}\mu^{-1})^{\phi^{-1}}}{\Gamma(\phi^{-1})}
y^{\phi^{-1}-1} e^{\phi^{-1}\mu^{-1}y}
\]</span></p>
<p>Where for <span class="math inline">\(y =0,1,2,...\)</span> the <span
class="math inline">\(E[Y] = \mu\)</span> and <span
class="math inline">\(Var[Y] = \mu^2 \phi\)</span>.</p>
</div>
<div id="simulation-parameters" class="section level3">
<h3><strong><em>Simulation Parameters</em></strong></h3>
<p>We will first simulate <span class="math inline">\(\lambda_j\)</span>
using the <strong>rgamma(</strong><span
class="math inline">\(n\)</span>,<span
class="math inline">\(\alpha\)</span>,<span
class="math inline">\(\sigma\)</span>) function in R with mean <span
class="math inline">\(\alpha\sigma\)</span> and variance <span
class="math inline">\(\alpha\sigma^2\)</span>. Thus, <span
class="math inline">\(\mu=\alpha\sigma\)</span> and <span
class="math inline">\(\phi = 1/\alpha\)</span>. We calculate <span
class="math inline">\(\mu_j\)</span> for each gene <span
class="math inline">\(j\)</span> and <span
class="math inline">\(L_i\)</span> for each cell <span
class="math inline">\(i\)</span> from the hybrid data set. For genes
with dispersion, the <span class="math inline">\(E[f] = \mu_0\)</span>.
We next simulate the Poisson distribution using the
<strong>rpois(</strong><span class="math inline">\(n\)</span>, <span
class="math inline">\(\lambda\)</span>) function in R. We simulate for
<span class="math inline">\(n=500\)</span> cells.</p>
</div>
</div>
</div>
<div id="simulations" class="section level1">
<h1><strong>Simulations</strong></h1>
<p>We first calculate the mean gene expression for in the hybrid dataset
using Human Cardiomyocytes from Replicate 3. Replicate 3 was chosen
since there are 546 cells, which is similar to the 500 cells we will
simulate.</p>
<pre class="r"><code>require(Seurat)
require(SeuratDisk)
require(SeuratData)
library(ggplot2)
library(cowplot)
library(scTenifoldNet)
HC &lt;- LoadH5Seurat(&quot;/project2/gilad/awchen55/differentialDispersion/data/hybrid_lines_raw_data/human.ASE.Rep3.h5Seurat&quot;)
cardiomyocytes_data &lt;- subset(HC, subset = (labels == &quot;Cardiomyocytes&quot;))
cardiomyocytes_expression_data &lt;- t(cardiomyocytes_data[[&#39;RNA&#39;]]$counts)</code></pre>
<p>We filter for low expression genes and calculate the <span
class="math inline">\(L_i\)</span> and relative expression to get <span
class="math inline">\(\lambda_j\)</span> values.</p>
<pre class="r"><code># calculate mean expression and remove genes with low expression
cardiomyocytes_expression_data_filtered &lt;- cardiomyocytes_expression_data[, colSums(cardiomyocytes_expression_data != 0) &gt; 50]

# calculate library size
library_sizes &lt;- rowSums(cardiomyocytes_expression_data_filtered)

# calculate the relative expression for each cell (quantity/rowsum)
relative_expression &lt;- t(apply(cardiomyocytes_expression_data_filtered,1, function(x) x/sum(x)))

# calculate mean relative expression per gene
mean_relative_expression &lt;- colMeans(relative_expression)

# calculate variance relative expression per gene
var_relative_expression &lt;-c()
for(i in 1:length(mean_relative_expression)){
  var_relative_expression &lt;- c(var_relative_expression,var(relative_expression[,i]))
}</code></pre>
<div id="mean-estimation" class="section level2">
<h2><strong>Mean Estimation</strong></h2>
<p>We calculate the sample mean of genes to get an idea of ranges of
mean expression. This will inform us of how to bin mean expression.</p>
<pre class="r"><code>hist(mean_relative_expression, breaks=100)</code></pre>
<p><img src="figure/simulation_null_analysis.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-5-1">
Past versions of unnamed-chunk-5-1.png
</button>
</p>
<div id="fig-unnamed-chunk-5-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/23e157da6638082dc5079827509d1f546ebfe4e9/docs/figure/simulation_null_analysis.Rmd/unnamed-chunk-5-1.png" target="_blank">23e157d</a>
</td>
<td>
awchen55
</td>
<td>
2024-08-13
</td>
</tr>
<tr>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/d68bca91fa39da29e9a25fa0f2cccc87a740dbc0/docs/figure/simulation_null_analysis.Rmd/unnamed-chunk-5-1.png" target="_blank">d68bca9</a>
</td>
<td>
awchen55
</td>
<td>
2024-08-06
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code># get the intervals of relative mean expression
mean_intervals &lt;- seq(min(mean_relative_expression), max(mean_relative_expression),length.out=101)</code></pre>
</div>
<div id="determine-dispersion-levels" class="section level2">
<h2><strong>Determine Dispersion Levels</strong></h2>
<p>We simulate a single gene for <span
class="math inline">\(n=500\)</span> cells and change the multiplier of
the <span class="math inline">\(\sigma\)</span> parameter. Again for the
<strong>rgamma</strong> function, the mean is <span
class="math inline">\(\alpha\sigma\)</span> and the variance is <span
class="math inline">\(\alpha \sigma^2\)</span>. Thus, we test different
multipliers for sigma to increase the variance in the simulated
dispersed genes.</p>
<pre class="r"><code># function to simulate a null gene and dispersed gene
sim_gene &lt;- function(n,alpha,sigma,Li, sigma_multiplier=1){

  # Simulate a single gene with following parameters:
  # n:                number of cells
  # alpha:            alpha parameter
  # sigma:            sigma parameter
  # Li:               library size
  # sigma_multiplier: factor to increase variance
 
  # Returns expression values for single gene in n cells
  
  # initialize Y
  Y &lt;- rep(NA,n) 
  
  # calculate latent expression using gamma distribution
  lambdaj &lt;- rgamma(n,shape = alpha,scale = sigma*sigma_multiplier)
  
  # for each cell use the latent expression and library size to simulate expression
  for(i in 1:n){
  
    Y[i] &lt;- rpois(1,Li*lambdaj[i])
  }
  
  return(Y)
}</code></pre>
<pre class="r"><code>set.seed(8626)
n=500
alpha = 0.5
sigma = 0.0002
Li = 8000

gene1 &lt;- sim_gene(n,alpha,sigma,Li)
gene1.5 &lt;- sim_gene(n,alpha,sigma,Li,1.5)
gene2 &lt;- sim_gene(n,alpha,sigma,Li,2)
gene2.5 &lt;- sim_gene(n,alpha,sigma,Li,2.5)
gene3 &lt;- sim_gene(n,alpha,sigma,Li,3)

df1 = rbind(data.frame(factor=&#39;1&#39;, UMI=gene1), data.frame(factor=&#39;1.5&#39;, UMI=gene1.5), data.frame(factor=&#39;2&#39;, UMI=gene2), data.frame(factor=&#39;2.5&#39;, UMI=gene2.5), data.frame(factor=&#39;3&#39;, UMI=gene3))

group.colors &lt;- c(&quot;red&quot;,&quot;blue&quot;,&quot;forestgreen&quot;,&quot;orange&quot;, &quot;purple&quot;)

ggplot(df1, aes(UMI, fill=factor)) + 
  geom_histogram( alpha=1, position=&quot;dodge&quot;,binwidth=0.5) + scale_fill_manual(values=group.colors) + scale_color_manual(values=group.colors) + theme(legend.key.size = unit(0.5, &#39;cm&#39;), legend.title = element_text(size=10)) + theme_bw() + xlim(-0.5, 10) + ggtitle(&quot;Histogram with Varying Dispersion for Low Alpha&quot;) </code></pre>
<p><img src="figure/simulation_null_analysis.Rmd/Dispersion%20Histogram%20Low%20Alpha-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-no-spaces-Dispersion-Histogram-Low-Alpha-1">
Past versions of "Dispersion Histogram Low Alpha-1.png"
</button>
</p>
<div id="fig-no-spaces-Dispersion-Histogram-Low-Alpha-1"
class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/f599f8dbcff6e9f62f33737f800434a72bb7b68b/docs/figure/simulation_null_analysis.Rmd/Dispersion Histogram Low Alpha-1.png" target="_blank">f599f8d</a>
</td>
<td>
awchen55
</td>
<td>
2024-08-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>set.seed(8626)
n=500
alpha = 5
sigma = 0.0002
Li = 8000

gene1 &lt;- sim_gene(n,alpha,sigma,Li)
gene1.5 &lt;- sim_gene(n,alpha,sigma,Li,1.5)
gene2 &lt;- sim_gene(n,alpha,sigma,Li,2)
gene2.5 &lt;- sim_gene(n,alpha,sigma,Li,2.5)
gene3 &lt;- sim_gene(n,alpha,sigma,Li,3)

df1 = rbind(data.frame(factor=&#39;1&#39;, UMI=gene1), data.frame(factor=&#39;1.5&#39;, UMI=gene1.5), data.frame(factor=&#39;2&#39;, UMI=gene2), data.frame(factor=&#39;2.5&#39;, UMI=gene2.5), data.frame(factor=&#39;3&#39;, UMI=gene3))

group.colors &lt;- c(&quot;red&quot;,&quot;blue&quot;,&quot;forestgreen&quot;,&quot;orange&quot;, &quot;purple&quot;)

ggplot(df1, aes(UMI, fill=factor)) + 
  geom_histogram( alpha=1, position=&quot;dodge&quot;,binwidth=1) + scale_fill_manual(values=group.colors) + scale_color_manual(values=group.colors) + theme(legend.key.size = unit(0.5, &#39;cm&#39;), legend.title = element_text(size=10)) + theme_bw() + ggtitle(&quot;Histogram with Varying Dispersion for High Alpha&quot;) + xlim(-0.5, 40) </code></pre>
<p><img src="figure/simulation_null_analysis.Rmd/Dispersion%20Histogram%20High%20Alpha-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-no-spaces-Dispersion-Histogram-High-Alpha-1">
Past versions of "Dispersion Histogram High Alpha-1.png"
</button>
</p>
<div id="fig-no-spaces-Dispersion-Histogram-High-Alpha-1"
class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/awchen55/differentialDispersion/blob/f599f8dbcff6e9f62f33737f800434a72bb7b68b/docs/figure/simulation_null_analysis.Rmd/Dispersion Histogram High Alpha-1.png" target="_blank">f599f8d</a>
</td>
<td>
awchen55
</td>
<td>
2024-08-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="simulating-for-bins-of-mean-expression" class="section level2">
<h2><strong>Simulating for Bins of Mean Expression</strong></h2>
<p>We have broken down the range of mean expression into 100 bins. For
each bin we will simulate 100 genes. 5% of those genes will have <span
class="math inline">\(\sigma\)</span> multiplied by a factor of 2. This
will yield 10,000 simulated genes with 5% of genes being significantly
dispersed.</p>
<pre class="r"><code># calculate alpha given mean expression and fixed sigma
calc_alpha &lt;- function(mu, sigma){
  alpha = mu/sigma
  return(alpha)
}


# Function to simulate genes within a bin of mean expression
sim_bin &lt;- function(n,p,sigma,sigma_multiplier,Li,mean_expr_lwr,mean_expr_upr){
  
  # Simulate all genes with following parameters:
  # n:                number of cells
  # p:                number of genes
  # sigma:            sigma parameter
  # sigma_multiplier: factor to increase variance
  # Li:               library size
  # mean_expr_lwr:    lower bound of mean expression window
  # mean_expr_upr:    upper bound of mean expression window
 
  # Returns matrix of expression for n cells and p genes. 
  # Returns vector of index values for simulated dispersed genes.
  
  # initialize Y for bin of expression
  Y_bin &lt;- matrix(NA,nrow=n,ncol=p)
  
  # breakdown expression range into p values
  bin_expression &lt;- seq(mean_expr_lwr,mean_expr_upr,length.out=p)
  
  # randomly select p expression means
  bin_sim_idx &lt;- sample.int(p, p, replace = TRUE)
  # randomly select 5% to be significantly dispersed
  bin_disp_idx &lt;- sample.int(p, 0.05*p, replace = FALSE)
  # get expression means
  sim_expression_values &lt;- bin_expression[bin_sim_idx]
  
  # simulate genes for each expression mean
  for(i in 1:p){
    
    # get expression mean
    mu &lt;- sim_expression_values[i]
    # calculate alpha for fixed sigma based on expression mean
    alpha &lt;- calc_alpha(mu,sigma)
    
    # for dispersed genes multiply sigma by factor
    if(i %in% bin_disp_idx){
      Y_bin[,i] &lt;- sim_gene(n,alpha,sigma,Li,sigma_multiplier)
    }
    else{
      Y_bin[,i] &lt;- sim_gene(n,alpha,sigma,Li)
    }
  }
  
  return(list(&quot;Y_bin&quot;=Y_bin, &quot;disp_idx&quot;=bin_disp_idx))
}

sim_Y &lt;- function(n,p,sigma,sigma_multiplier,Li,mean_expr_lwr,mean_expr_upr){
  Y &lt;- c()
  disp_idx &lt;- c()
  for(i in 1:100){
    
    mean_expr_lwr &lt;- mean_intervals[i]
    mean_expr_upr &lt;- mean_intervals[i+1]
    
    gene_sim &lt;- sim_bin(n,p,sigma,sigma_multiplier,Li,mean_expr_lwr,mean_expr_upr)
    Y_bin &lt;- gene_sim$Y_bin
    bin_disp_idx &lt;- gene_sim$disp_idx + (i-1)*100
    
    Y &lt;- cbind(Y,Y_bin)
    disp_idx &lt;- c(disp_idx,bin_disp_idx)
  }
  return(list(&quot;Y&quot;=Y,&quot;disp_idx&quot;=disp_idx))
}


cpm_norm &lt;- function(Y){
  
  require(scTenifoldNet)
  Y_cpm_normalized &lt;- t(cpmNormalization(t(Y)))
  Y_log_cpm_normalized &lt;- log(t(cpmNormalization(t(Y))))
  
  Y_cv &lt;- apply(Y_cpm_normalized, 2, sd)/apply(Y_cpm_normalized, 2, mean)
  Y_mean &lt;- apply(Y_cpm_normalized, 2, mean)
  Y_cv_log &lt;- apply(Y_log_cpm_normalized, 2, sd)/apply(Y_log_cpm_normalized, 2, mean)
  Y_mean_log &lt;- apply(Y_log_cpm_normalized, 2, mean)
  
  return(list(&quot;Y_cpm_normalized&quot;=Y_cpm_normalized, &quot;Y_log_cpm_normalized&quot; = Y_log_cpm_normalized, &quot;Y_cv&quot; = Y_cv, &quot;Y_mean&quot; = Y_mean, &quot;Y_cv_log&quot;=Y_cv_log, &quot;Y_mean_log&quot; = Y_mean_log))
}</code></pre>
<p>We simulate applying different factors to amplify the variance.</p>
<pre class="r"><code>#### Simulate Y matrix for all genes
# Parameters
n=500
p=100
sigma = 0.0002
Li = 8000

Y_1.25 &lt;- sim_Y(n,p,sigma,1.25,Li,mean_expr_lwr,mean_expr_upr)
Y_1.5 &lt;- sim_Y(n,p,sigma,1.5,Li,mean_expr_lwr,mean_expr_upr)
Y_1.75 &lt;- sim_Y(n,p,sigma,1.75,Li,mean_expr_lwr,mean_expr_upr)
Y_2 &lt;- sim_Y(n,p,sigma,2,Li,mean_expr_lwr,mean_expr_upr)</code></pre>
<p>Perform normalization on data.</p>
<pre class="r"><code>Y_1.25_cpm &lt;- cpm_norm(Y_1.25$Y)
Y_1.5_cpm &lt;- cpm_norm(Y_1.5$Y)
Y_1.75_cpm &lt;- cpm_norm(Y_1.75$Y)
Y_2_cpm &lt;- cpm_norm(Y_2$Y)


Y_1.25_cv &lt;- apply(Y_1.25$Y, 2, sd)/apply(Y_1.25$Y, 2, mean)
Y_1.25_mean &lt;- apply(Y_1.25$Y, 2, mean)

Y_1.5_cv &lt;- apply(Y_1.5$Y, 2, sd)/apply(Y_1.5$Y, 2, mean)
Y_1.5_mean &lt;- apply(Y_1.5$Y, 2, mean)

Y_1.75_cv &lt;- apply(Y_1.75$Y, 2, sd)/apply(Y_1.75$Y, 2, mean)
Y_1.75_mean &lt;- apply(Y_1.75$Y, 2, mean)

Y_2_cv &lt;- apply(Y_2$Y, 2, sd)/apply(Y_2$Y, 2, mean)
Y_2_mean &lt;- apply(Y_2$Y, 2, mean)</code></pre>
<p>We compare CPM vs log(CPM) normalization. We see the increase effect
of the factor. For high expressed genes, we see the mean-variance
relationship, which may make it difficult to identify signifcantly
dispersed genes.</p>
<pre class="r"><code>par(mfrow = c(2, 2))

plot(log(Y_1.25_cpm$Y_mean),Y_1.25_cpm$Y_cv, pch=16, cex=0.5 ,main = &quot;Factor: 1.25&quot;, xlab=&quot;log(mean)&quot;, ylab=&quot;CV&quot;)
points(log(Y_1.25_cpm$Y_mean[Y_1.25$disp_idx]),Y_1.25_cpm$Y_cv[Y_1.25$disp_idx],pch=16,col=&quot;red&quot;,cex=0.5)

plot(log(Y_1.5_cpm$Y_mean),Y_1.5_cpm$Y_cv, pch=16, cex=0.5 ,main = &quot;Factor: 1.5&quot;, xlab=&quot;log(mean)&quot;, ylab=&quot;CV&quot;)
points(log(Y_1.5_cpm$Y_mean[Y_1.5$disp_idx]),Y_1.5_cpm$Y_cv[Y_1.5$disp_idx],pch=16,col=&quot;red&quot;,cex=0.5)

plot(log(Y_1.75_cpm$Y_mean),Y_1.75_cpm$Y_cv, pch=16, cex=0.5 ,main = &quot;Factor: 1.75&quot;, xlab=&quot;log(mean)&quot;, ylab=&quot;CV&quot;)
points(log(Y_1.75_cpm$Y_mean[Y_1.75$disp_idx]),Y_1.75_cpm$Y_cv[Y_1.75$disp_idx],pch=16,col=&quot;red&quot;,cex=0.5)

plot(log(Y_2_cpm$Y_mean),Y_2_cpm$Y_cv, pch=16, cex=0.5 ,main = &quot;Factor: 2&quot;, xlab=&quot;log(mean)&quot;, ylab=&quot;CV&quot;)
points(log(Y_2_cpm$Y_mean[Y_2$disp_idx]),Y_2_cpm$Y_cv[Y_2$disp_idx],pch=16,col=&quot;red&quot;,cex=0.5)
mtext(&quot;CPM Normalized&quot;, side = 3, line = - 1, outer = TRUE)</code></pre>
<p><img src="figure/simulation_null_analysis.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>par(mfrow = c(2, 2))

plot(Y_1.25_cpm$Y_mean_log,Y_1.25_cpm$Y_cv_log, pch=16, cex=0.5 ,main = &quot;Factor: 1.25&quot;, xlab=&quot;log(mean)&quot;, ylab=&quot;CV&quot;)
points(Y_1.25_cpm$Y_mean_log[Y_1.25$disp_idx],Y_1.25_cpm$Y_cv_log[Y_1.25$disp_idx],pch=16,col=&quot;red&quot;,cex=0.5)

plot(Y_1.5_cpm$Y_mean_log,Y_1.5_cpm$Y_cv_log, pch=16, cex=0.5 ,main = &quot;Factor: 1.5&quot;, xlab=&quot;log(mean)&quot;, ylab=&quot;CV&quot;)
points(Y_1.5_cpm$Y_mean_log[Y_1.5$disp_idx],Y_1.5_cpm$Y_cv_log[Y_1.5$disp_idx],pch=16,col=&quot;red&quot;,cex=0.5)

plot(Y_1.75_cpm$Y_mean_log,Y_1.75_cpm$Y_cv_log, pch=16, cex=0.5 ,main = &quot;Factor: 1.75&quot;, xlab=&quot;log(mean)&quot;, ylab=&quot;CV&quot;)
points(Y_1.75_cpm$Y_mean_log[Y_1.75$disp_idx],Y_1.75_cpm$Y_cv_log[Y_1.75$disp_idx],pch=16,col=&quot;red&quot;,cex=0.5)

plot(Y_2_cpm$Y_mean_log,Y_2_cpm$Y_cv_log, pch=16, cex=0.5 ,main = &quot;Factor: 2&quot;, xlab=&quot;log(mean)&quot;, ylab=&quot;CV&quot;)
points(Y_2_cpm$Y_mean_log[Y_2$disp_idx],Y_2_cpm$Y_cv_log[Y_2$disp_idx],pch=16,col=&quot;red&quot;,cex=0.5)

mtext(&quot;Log CPM Normalized&quot;, side = 3, line = - 1, outer = TRUE)</code></pre>
<p><img src="figure/simulation_null_analysis.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.0 (2022-04-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.3.13-el7-x86_64/lib/libopenblas_haswellp-r0.3.13.so

locale:
 [1] LC_CTYPE=en_US.UTF-8 LC_NUMERIC=C         LC_TIME=C           
 [4] LC_COLLATE=C         LC_MONETARY=C        LC_MESSAGES=C       
 [7] LC_PAPER=C           LC_NAME=C            LC_ADDRESS=C        
[10] LC_TELEPHONE=C       LC_MEASUREMENT=C     LC_IDENTIFICATION=C 

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] scTenifoldNet_1.3     cowplot_1.1.1         ggplot2_3.5.1        
[4] SeuratData_0.2.2.9001 SeuratDisk_0.0.0.9021 Seurat_5.1.0         
[7] SeuratObject_5.0.2    sp_1.6-0             

loaded via a namespace (and not attached):
  [1] spam_2.8-0             workflowr_1.7.1        plyr_1.8.7            
  [4] igraph_2.0.3           lazyeval_0.2.2         splines_4.2.0         
  [7] RcppHNSW_0.4.1         listenv_0.8.0          scattermore_1.2       
 [10] digest_0.6.29          htmltools_0.5.2        fansi_1.0.3           
 [13] magrittr_2.0.3         tensor_1.5             cluster_2.1.3         
 [16] ROCR_1.0-11            globals_0.15.0         matrixStats_0.62.0    
 [19] spatstat.sparse_3.0-0  colorspace_2.0-3       rappdirs_0.3.3        
 [22] ggrepel_0.9.5          xfun_0.30              dplyr_1.1.4           
 [25] crayon_1.5.1           jsonlite_1.8.0         progressr_0.10.0      
 [28] spatstat.data_3.0-0    survival_3.3-1         zoo_1.8-10            
 [31] glue_1.7.0             polyclip_1.10-0        gtable_0.3.0          
 [34] leiden_0.4.2           future.apply_1.9.0     abind_1.4-5           
 [37] scales_1.3.0           DBI_1.1.2              spatstat.random_3.1-3 
 [40] miniUI_0.1.1.1         Rcpp_1.0.12            viridisLite_0.4.0     
 [43] xtable_1.8-4           reticulate_1.24        bit_4.0.4             
 [46] dotCall64_1.0-1        htmlwidgets_1.5.4      httr_1.4.3            
 [49] RColorBrewer_1.1-3     ellipsis_0.3.2         ica_1.0-2             
 [52] farver_2.1.0           pkgconfig_2.0.3        sass_0.4.1            
 [55] uwot_0.1.14            deldir_1.0-6           utf8_1.2.2            
 [58] tidyselect_1.2.1       labeling_0.4.2         rlang_1.1.4           
 [61] reshape2_1.4.4         later_1.3.0            munsell_0.5.0         
 [64] tools_4.2.0            cli_3.6.3              generics_0.1.3        
 [67] ggridges_0.5.3         evaluate_0.15          stringr_1.5.1         
 [70] fastmap_1.1.0          yaml_2.3.5             goftest_1.2-3         
 [73] RhpcBLASctl_0.23-42    knitr_1.39             bit64_4.0.5           
 [76] fs_1.5.2               fitdistrplus_1.1-8     purrr_1.0.2           
 [79] RANN_2.6.1             pbapply_1.5-0          future_1.25.0         
 [82] nlme_3.1-157           whisker_0.4            mime_0.12             
 [85] hdf5r_1.3.8            compiler_4.2.0         rstudioapi_0.13       
 [88] plotly_4.10.0          png_0.1-7              spatstat.utils_3.0-1  
 [91] tibble_3.2.1           bslib_0.3.1            stringi_1.7.6         
 [94] highr_0.9              RSpectra_0.16-1        lattice_0.20-45       
 [97] Matrix_1.6-4           vctrs_0.6.5            pillar_1.9.0          
[100] lifecycle_1.0.4        spatstat.geom_3.0-6    lmtest_0.9-40         
[103] jquerylib_0.1.4        RcppAnnoy_0.0.19       data.table_1.14.2     
[106] irlba_2.3.5.1          httpuv_1.6.5           patchwork_1.2.0       
[109] R6_2.5.1               promises_1.2.0.1       KernSmooth_2.23-20    
[112] gridExtra_2.3          parallelly_1.31.1      codetools_0.2-18      
[115] fastDummies_1.7.3      MASS_7.3-56            rprojroot_2.0.3       
[118] withr_3.0.0            sctransform_0.4.1      parallel_4.2.0        
[121] grid_4.2.0             tidyr_1.3.1            rmarkdown_2.14        
[124] Rtsne_0.16             git2r_0.30.1           spatstat.explore_3.0-6
[127] shiny_1.7.1           </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
