---
title: "**Poisson and Negative Binomial Simulations**"
output: html_document
date: '2024-07-24'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# **Simulation Details**

We simulate a matrix $Y \in \mathbb{R}^{n\times p}$ that contains raw expression values (scRNA-seq counts) for $p$ genes and $n$ single cells. We set $\pi$ as the percent of genes that are overdispersed. We use a Gamma-Poisson model for each gene and can set the dispersion level for each simulated gene.

```{r, eval=F,include=F}
## **Poisson Distribution**

### ***Model***

We first simulate a Poisson distribution as our null model for each gene, where there is no dispersion parameter.

$$
y_{ij} \sim Pois(L_i\lambda_j) 
$$

Where the variables for cell $i$ and gene $j$ are:

-   $y_{ij}$ is the number of unique molecular identifier (UMI) for cell $i$ and gene $j$

-   $L_i$ is the library size of cell $i$

-   $\lambda_j$ is the relative expression value of gene $j$

### ***Density Function***

The density function for the Poisson distribution is:

$$
f(y;\lambda) = \frac{\lambda^y e^{-\lambda}}{y!}
$$

Where for $y =0,1,2,...$ the $E[Y] = Var[Y] = \lambda$.

### ***Simulation Parameters***

We use the **rpois(**$n$, $\lambda$) function in R to simulate the Poisson distribution. The library size and relative expression values are calculated from a real data set (human-chimp hybrid data). Each $Y$ is calculated for cells from the same species, cell type, and replicate. We simulate for $n=500$ cells.
```

## **Gamma-Poisson (NB) Distribution**

### ***Model***

We simulate the Negative Binomial distribution using the Gamma-Poisson distribution, since the Gamma distribution is a conjugate prior of the Poisson distribution. Under this model,

$$
y_{ij} \sim Pois(L_i \lambda_j)
$$

$$
\lambda_j \sim \begin{equation}
  f(\cdot) =
    \begin{cases}
      \pi & Gamma(\mu_j,\phi_j)\\
      (1-\pi) & Gamma(\mu_0,\phi_0) \\
    \end{cases}       
\end{equation}
$$

Where the variables for cell $i$ and gene $j$ are:

-   $y_{ij}$ is the number of unique molecular identifier (UMI) for cell $i$ and gene $j$

-   $L_i$ is the library size of cell $i$

-   $\lambda_j$ is the latent expression value of gene $j$

-   $\pi$ is the percent of genes that are overdispersed

-   $\mu_j$ is the mean relative expression value of $\lambda_j$ for gene $j$

-   $\mu_0$ is the null relative expression value for a gene that has no dispersion

-   $\phi_j$ is the dispersion value of $\lambda_j$ for gene $j$

-   $\phi_0$ is the null relative dispersion value for a gene that has no dispersion

### ***Density Function***

The density function for the Gamma distribution is:

$$
f(y; \mu,\phi) = \frac{(\phi^{-1}\mu^{-1})^{\phi^{-1}}}{\Gamma(\phi^{-1})} y^{\phi^{-1}-1} e^{\phi^{-1}\mu^{-1}y}
$$

Where for $y =0,1,2,...$ the $E[Y] = \mu$ and $Var[Y] = \mu^2 \phi$.

### ***Simulation Parameters***

We will first simulate $\lambda_j$ using the **rgamma(**$n$,$\alpha$,$\sigma$) function in R with mean $\alpha\sigma$ and variance $\alpha\sigma^2$. Thus, $\mu=\alpha\sigma$ and $\phi = 1/\alpha$. We calculate $\mu_j$ for each gene $j$ and $L_i$ for each cell $i$ from the hybrid data set. For genes with dispersion, the $E[f] = \mu_0$. We next simulate the Poisson distribution using the **rpois(**$n$, $\lambda$) function in R. We simulate for $n=500$ cells.

```{r, eval=F,include=F}
## **Negative Binomial Distribution**

We simulate the Negative Binomial distribution as the the alternative model for each gene. The density function is:

$$
f(y;\mu, \theta) = {y+\theta-1 \choose y} \left(\frac{\mu}{\mu+\theta}\right)^y \left(\frac{\theta}{\mu+\theta}\right)^\theta 
$$ where for $y =0,1,2,...$ the $E[Y] = \mu$ and $Var[Y] = \mu + \frac{\mu^2}{\theta}$. Note that the *dispersion* parameter is $\phi=1/\theta$. We use the **rnegbin(**$n$, $\mu$, $\theta$) function from the *MASS* R package to simulate the Negative Binomial distribution.

$$
y_{ij} \sim NB(\mu_j,\theta_j) \text{ ,    where } j = 1,\cdots,p
$$

We set $\mu_j$ equal to the mean gene expression using raw counts from the hybrid data set. Mean gene expression is calculated across cells from the same species, replicate, and cell type. We introduce and test varying dispersion levels for $\phi_j = 1/\theta_j$ and simulate for $n=500$ cells.

```

# **Simulations**

We first calculate the mean gene expression for in the hybrid dataset using Human Cardiomyocytes from Replicate 3. Replicate 3 was chosen since there are 546 cells, which is similar to the 500 cells we will simulate.

```{r, message=F, warning=F}
require(Seurat)
require(SeuratDisk)
require(SeuratData)
HC <- LoadH5Seurat("/project2/gilad/awchen55/differentialDispersion/data/hybrid_lines_raw_data/human.ASE.Rep3.h5Seurat")
cardiomyocytes_data <- subset(HC, subset = (labels == "Cardiomyocytes"))
cardiomyocytes_expression_data <- t(cardiomyocytes_data[['RNA']]$counts)
```

We filter for low expression genes and calculate the $L_i$ and relative expression to get $\lambda_j$ values.

```{r}
# calculate mean expression and remove genes with low expression
cardiomyocytes_expression_data_filtered <- cardiomyocytes_expression_data[, colSums(cardiomyocytes_expression_data != 0) > 50]

# calculate library size
library_sizes <- rowSums(cardiomyocytes_expression_data_filtered)

# calculate the relative expression for each cell (quantity/rowsum)
relative_expression <- t(apply(cardiomyocytes_expression_data_filtered,1, function(x) x/sum(x)))

# calculate mean relative expression per gene
mean_relative_expression <- colMeans(relative_expression)

# calculate variance relative expression per gene
var_relative_expression <-c()
for(i in 1:length(mean_relative_expression)){
  var_relative_expression <- c(var_relative_expression,var(relative_expression[,i]))
}
```

We estimate the mean and variance from the relative expression to calculate estimates for $\alpha = mean^2/variance$ and $\sigma = variance/mean$.

```{r}
alphas <- mean_relative_expression^2/var_relative_expression
sigmas <- var_relative_expression/mean_relative_expression
```

We use the estimates for $\alpha$ and $\sigma$ to simulate the $\lambda_j \sim Gamma(\alpha,\sigma)$. The simulated $\lambda_j$ and library size $L_i$ is used to simulate $y_{ij} \sim Pois(L_i\lambda_j)$. We do this first for 1 example gene.

```{r}
set.seed(8626)
n=500
gamma_dist <- rgamma(n,shape = alphas[1],scale = sigmas[1])
library_size_lambda <- library_sizes[1]*gamma_dist

yi <- c()
for(i in 1:n){
  yi <- c(yi,rpois(1,library_size_lambda[1]))
}

```

We compare the histograms of UMI in 1 real versus simulated gene.

```{r}
library(ggplot2)
df1 = rbind(data.frame(grp='real data', UMI=cardiomyocytes_expression_data_filtered[,1]), data.frame(grp='simulated', UMI=yi))
group.colors <- c("red","blue")
ggplot(df1, aes(UMI, fill=grp)) + 
  geom_histogram(color='black', alpha=1, position="dodge",binwidth=0.5) + scale_fill_manual(values=group.colors)
```

We now simulate for all genes. We set $\pi=5%$ of genes to be overdispersed.

```{r}
pi_disp = 0.05
n=500
p = 1000#ncol(cardiomyocytes_expression_data_filtered)
dispersed_idx = sample(p,pi_disp*p)
Y <- matrix(nrow=n,ncol=p)

mean_sim <- c()
var_sim <- c()

for(j in 1:p){

  if(i %in% dispersed_idx){
    lambdaj <- rgamma(n,shape = mean(alphas)/10,scale = means(sigmas))
  }
  else{
    lambdaj <- rgamma(n,shape = mean(alphas),scale = mean(sigmas))
  }

  for(i in 1:n){
    Y[i,j] = rpois(1,library_sizes[i]*lambdaj[i])
  }

  mean_sim <- c(mean_sim,mean(Y[,j]))
  var_sim <- c(var_sim, var(Y[,j]))
}

```


```{r}

plot(mean_sim,var_sim,pch=16,col="blue")
points(mean_sim[dispersed_idx],var_sim[dispersed_idx],pch=16,col="red")
```

```{r,eval=F,include=F}
m2 <- glm(relative_expression[,1]~1, family = Gamma(link = "identity"))
summary(m2)
```
